runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java

```diff
@@ -201,56 +201,66 @@ public int hashCode() {
    */
   public class Reader extends BoundedSource.BoundedReader<T> {
     private long recordsRead = 0L;
-    private Instant endTime;
+    private Instant readEndTime;
     private final FluentBackoff backoffFactory;
-    private final UnboundedSource.UnboundedReader<T> reader;
+    private final UnboundedSource.UnboundedReader<T> unboundedReader;
     private boolean started;
 
-    private Reader(UnboundedSource.UnboundedReader<T> reader) {
-      this.reader = reader;
+    private Reader(final UnboundedSource.UnboundedReader<T> unboundedReader) {
+      this.unboundedReader = unboundedReader;
       backoffFactory =
           FluentBackoff.DEFAULT
               .withInitialBackoff(Duration.millis(10))
               .withMaxBackoff(maxReadTime.minus(1))
               .withMaxCumulativeBackoff(maxReadTime.minus(1));
     }
 
+    private boolean startIfNeeded() throws IOException {
+      return !started && ((started = true) && unboundedReader.start());
```

Comment:
Consider the following snippet [taken from here](https://github.com/apache/beam/blob/master/runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/StateSpecFunctions.java#L168-L173):

```java
          boolean finished = !microbatchReader.start();
          while (!finished) {
            final WindowedValue<T> wv =
                WindowedValue.of((T) microbatchReader.getCurrent(),
                                 microbatchReader.getCurrentTimestamp(),
                                 GlobalWindow.INSTANCE,
                                 PaneInfo.NO_FIRING);
            readValues.add(CoderHelpers.toByteArray(wv, coder));
            finished = !microbatchReader.advance();
          }
```

`recordsRead` lags by 1, and essentially counts the records that have already been added to `readValues`. 

After `start()` is called for the first time, it advances the reader to the first record, and the corresponding `recordsRead` update takes place in `microbatchReader.advance()` at the end of the `while` loop.